### PR TITLE
Milestone 3: Friendship v2

### PR DESCRIPTION
**Pull request by fernandorpm**
# Milestone 3: Friendship v2

## To update
- friendship table, model, views
- user model, views
- unfriend option, reject friendship request

### CODE REVIEW

**Comment by fernandorpm** [comment]
## To the code reviewer
The Milestone tells us to implement mutual friendships, which my project already does, so I'm unsure if there are any necessary changes. Any advise is appreciated, so please, let me know if you have some! :rocket:

**Review by elisha2kyakpo1 — CHANGES_REQUESTED** [review]
# Status: Changes required👈
Hey, you have done a good job and I'm happy for you. However, we have to make some changes before the project gets approved.

<hr>
**User model**

- [x] A user should have many pending friend requests waiting for acceptance, like has_many :pending_friendships.
- [x]  A user should have many friends, like has_many :friends. 
- [x] in your model you have used receiver and sender, does this represent received and sent requests? if yes, then you should have a has_many :confirmed_friendships, etc.
- [x] Try to give meaningful names like received_requests, sent_friend_requests instead of has_many :senders or receivers, for a user.
- [x] when friendship is confirmed, an additional row with reversed user-friend pair is created, something like this; 
```ruby
def confirm_friend
  self.update_attributes(confirmed: true)
  Friendship.create!(friend_id: self.user_id,
                  user_id: self.friend_id,
                  confirmed: true)
end
```
- [x] You should go through [this artcile](https://explainextended.com/2009/03/07/selecting-friends) and [stackoverflow discussion](https://stackoverflow.com/questions/4219979/sql-best-practice-for-a-friendship-table) as a hint.

<hr>
In case you have a question, please reach me on slack as Elisha kyakopo
Cheers💪🎉

File: app/models/friendship.rb
Comment (rindrajosia) [inline app/models/friendship.rb:7]: Please check if the friendship request was accepted

**Review by rindrajosia — CHANGES_REQUESTED** [review]
Changes Required :red_circle:
To Highlight: :angel:

- Good job by trying to add row with reversed user-friend pair. :+1: 

To improve :factory_worker: [REQUIRED]:

- [x] Please, make sure that you check if the friendship is confirmed before creating the additional row with reversed user-friend pair instead of just to check if friendship exists.
- [x]  Thanks to this, we can find user friends only by user_id in friendships table
```
# User model

has_many :confirmed_friendships, -> { where accepted: true }, class_name: "Friendship"
has_many :friends, through: :confirmed_friendships
```

- [ ] User#pending_friends can be defined as association with `SQL query instead of processing objects in application memory` (**as the database was created to store and query data it is always more efficient than ruby**)
```
# User model

# Users who needs to confirm friendship
has_many :pending_friendships, -> { where accepted: false }, class_name: "Friendship", foreign_key: "user_id"
has_many :pending_friends, through: :pending_friendships, source: :friend
```

- [ ] User#friend_requests can be defined as association with `SQL query instead of processing objects in application memory` (**as the database was created to store and query data it is always more efficient than ruby**)
```
# User model

# Users who requested to be friends (needed for notifications)
has_many :inverted_friendships, -> { where confirmed: false }, class_name: "Friendship", foreign_key: "friend_id"
has_many :friend_requests, through: :inverted_friendships
```

- [x] in order to get all posts written by user on their friends, students can use simple method in User model (in order to **NOT** do it with Enumerables)
```
# User model

def friends_and_own_posts
  Post.where(user: (self.friends.to_a << self))
  # This will produce SQL query with IN. Something like: select * from posts where user_id IN (1,45,874,43);
end
```

Nice to have [OPTIONAL] :

- [x]  It will be nice if  there are no logic in your view and partial.
- [x] Please read about n+1 and make sure there are no n + 1 queries in index views [Recommended article.](https://medium.com/@bretdoucette/n-1-queries-and-how-to-avoid-them-a12f02345be5)

That's it, you can submit another code review request after making these changes :recycle: :innocent:
With all my respect :pray:
Rindra Josia

**Comment by fernandorpm** [comment]
# To the next reviewer

Some of the suggested changes does not align to what was proposed by the challenge, nor to my code.

For example, when the reviewer suggests to show `pending_friendships`, but mentioned before that the friendship records/rows should only be duplicated if they were accepted, he is contradicting himself.

What I have done is to change my code to replace the `pending_requests` filter as a query (as he suggested), in the `has_many` relationship (which I totally agree with!).

**Review by MkrtichSargsyan — CHANGES_REQUESTED** [review]
## _Status:_ Changes Requested :recycle:

## Hello Fernando Marques



THIS IS VERY IMPRESSIVE WORK. 
You have done good job guys, from the previous review  :+1:
However, there are some changes that you need to make in order to get your project approved

- [x] I got this error message when I clicked on the add friend button
err

Please, fix this error and submit your request again! :grinning:
Good luck. 

> ps: . If you have any questions about my review or need help, you can message me on slack ( @mkrtichsargsyan )

**Review by mke2111 — APPROVED** [review]
# 😄 🥇  STATUS: APPROVED  🥇 👏🏽 

## Hey Happy Coder, 👋🏽 
Congratulations, 🥂 . You got minimum requirements for approval. 👍🏽 

### **_Well Done_** 💯 

    1. The friend request invite functionality works just fine
    2. The linter checks pass successfully
    

### **_Room for Improvement_** 🤪 


- [x] It is a good practice to have zero logic in the views directory. Create methods in the helpers directory and call the methods in the views directory instead.
Read this [article](https://medium.com/@acushla4real/rails-view-helpers-best-practices-guide-d7d7b6db3db4) for a head-start. Another great piece to read is this [article](https://brandonhilkert.com/blog/refactoring-logic-from-a-rails-view/)

### You may now merge to the develop branch and remember to submit a project completion form.
### I wish you the very best in your upcoming projects.

**Cheers 🍹 and Happy coding 💻** 

<hr>

### Best regards
Slack 🆘  Roy Mukuye
Twitter 🐦  [Roy Mukuye](https://twitter.com/Roymkenya)
LinkedIn 🔗  [Roy Mukuye](https://www.linkedin.com/in/roy-mukuye/)

**Comment by rindrajosia** [comment]
> # To the next reviewer
> 
> Some of the suggested changes does not align to what was proposed by the challenge, nor to my code.
> 
> For example, when the reviewer suggests to show `pending_friendships`, but mentioned before that the friendship records/rows should only be duplicated if they were accepted, he is contradicting himself.
> 
> What I have done is to change my code to replace the `pending_requests` filter as a query (as he suggested), in the `has_many` relationship (which I totally agree with!).

Hello 
I am sorry to mislead you. I haven't done that on purpose :blush:

**Pull request merged**
